lambda/value_rec_compiler.ml

```diff
@@ -254,6 +312,63 @@ let lifted_block_mut : Asttypes.mutable_flag = Immutable
 
 let no_loc = Debuginfo.Scoped_location.Loc_unknown
 
+
+(** {1. Function Lifting} *)
+
+(* The compiler allows recursive definitions of functions that are not
+   syntactic functions:
+   {[
+     let rec f_syntactic_function = fun x ->
+       f_syntactic_function x
+
+     let rec g_needs_lift =
+       let () = ... in
+       (fun x -> g_needs_lift (foo x))
+
+     let rec h_needs_lift_and_closure =
+       let v = ref 0 in
+       (fun x -> incr v; h_needs_lift_and_closure (bar x))
+
+     let rec i_needs_lift_and_eta =
+       let aux x = i_needs_lift_and_eta (baz x) in
+       aux
+   ]}
+
+   We need to translate those using only syntactic functions or blocks.
+   For some functions, we only need to lift a syntactic function in tail
+   position from its surrounding context:
+   {[
+     let rec g_context =
+       let () = ... in
+       ()
+     and g_lifted = fun x ->
+       g_lifted (foo x)
+   ]}
+
+   In general the function may refer to local variables, so we perform
+   a local closure conversion before lifting:
+   {[
+     let rec h_context =
+       let v = ref 0 in
+       { v }
+     and h_lifted = fun x ->
+       incr h_context.v;
+       h_lifted (bar x)
+   ]}
+   Note that the closure environment computed from the context is passed as a
+   mutually recursive definition, that is, a free variable, and not as an
+   additional function parameter (which is customary for closure conversion).
+
+   Finally, when the tail expression is a variable, we perform an eta-expansion
+   to get a syntactic function, that we can then close and lift:
+   {[
+     let rec i_context =
+       let aux x = i_lifted (baz x) in
+       { aux }
+     and i_lifted = fun x -> i_context.aux x
+   ]}
+*)
```

Comment:
We forgot to say a word about whether this transformation is performance-preserving. My understanding is that closure conversion introduces a slight overhead as there will be an extra indirection to access the free local variables (the block is in the closure, and then we index into that block, the extra indexing is a new cost) -- whenever the function is called later. One should of course point out that syntactic functions are unchanged and this is the most common case. You could also mention that avoiding backpatching for functions lets the optimizer do a better job.

Reply:
Is there a compilation scheme that avoids the extra indirection, by defining each free variable separately in the recursive block? My understanding is that this is not easy to do in general, because the free variables may depend on each other in a way that invalidates the naive strategy of simply lifting each of them as one recursive definition. For example:

```ocaml
let rec f =
  let v0 = ... f ... in
  let v1 = ... f ... v0 ... in
  fun x -> ... v0 ... v1 ...
```

cannot always be turned into the obvious

```ocaml
let rec f = fun x -> ... v0 ... v1 ...
and v0 = ... f ...
and v1 = ... f ... v0 ...
```

as the dependency of `v1` over `v0` may be stricter than what is allowed between recursive neighbors.

Maybe we *could* do something by remembering an ordering constraint between `v0` and `v1` for the code generation pass later. This sounds complex for a worry that is mostly theoretical, so I think that the current choice is reasonable.

Reply:
This is the approach taken in the dissection PR. See ocaml-flambda/flambda-backend#1814 for an example where it doesn't work.

Reply:
(I would still welcome a short comment on the code-generation / performance impact of the transformation.)

Reply:
Yes, I haven't missed it and I'm working on it.

Reply:
This is actually something we discussed for Flambda 2 (in the context of regular closure conversion, but the issue is the same).
With big environments, by reading them early you risk having many variables live at the same time, making register allocation less efficient. But if you have several occurrences of the same variable, the multiple loads will add some extra instructions (increasing code size and the number of runtime instructions).
Closure always uses the substitution version, as does Flambda 1. Flambda 2 used to do the version with early bindings, but we had to change it towards a substitution-like version because of some bad cases.
So I think the version with substitutions is less likely to be very bad, but there are probably a number of common cases where it's not as good as early bindings.

Reply:
Interesting! I am glad that I asked.